src/library/scala/collection/Iterator.scala

```diff
@@ -590,14 +590,8 @@ trait Iterator[+A] extends IterableOnce[A] with IterableOnceOps[A, Iterator, Ite
     def next() = if (hasNext) { hdDefined = false; hd } else Iterator.empty.next()
   }
 
-  def drop(n: Int): Iterator[A] = {
-    var i = 0
-    while (i < n && hasNext) {
-      next()
-      i += 1
-    }
-    this
-  }
+  def drop(n: Int): Iterator[A] = sliceIterator(n, -1)
+
```

Comment:
It looks like `sliceIterator` also does the while loop calling `next()` n times, but allocates a new Iterator object, whereas the existing implementation doesn't allocate a new instance.

Somewhere where you could maybe improve things is on `IndexedSeqViewIterator`. `IndexedSeq`'s default iterators are formed by `this.view.iterator` which will form an `IndexedSeqViewIterator`, but that class does not overrride `drop` to take advantage of the fact that it is indexed and so can just update its position.

Reply:
Yes, I was aware that `sliceIterator` will take linear time when calling `next because of the skipping`. In the context of just the `drop` method (disregarding whether `next` will be called or not), it takes constant time, whereas previously it was linear time. 
Performance-wise using the iterator's `next` functionality, there won't be any significant performance improvement with this approach/code, but I assumed this was the intended design (making *just* the `drop` to have constant time) according to the post mentioning the `take` for `Iterator` should also take constant time.

Understood. @szeiger mentioned in the thread we can generalize `IndexedSeqViewIterator` to `IndexedIterator` that can be used with *any* `Iterable` such that the method `drop` and the use (next) can still satisfy constant time. Given that `LinearSeq` is an `Iterable`, I assume the optimizations required are different from just implementing `IndexedSeqViewIterator` drop operations? Will that be even possible to achieve constant time from the overall use of `drop` and `next` call in constant time for `Iterable` via `Iterator`? Or maybe I misunderstood his post.

Reply:
@joshlemer Also, the [drop](https://github.com/scala/scala/blob/ba9701059216c629410f4f23a2175d20ad62484b/src/library/scala/collection/IndexedSeqView.scala#L54) in `IndexedSeqViewIterator` iterator is already constant time.
It modifies the `current` variable, which is used to determine the next element's index.